File: dsa/examples/use-cases/document-signing/verify_document.py
#!/usr/bin/env python3
"""
Post-Quantum Document Signature Verification Tool

Verifies document signatures created with sign_document.py.

Usage:
    python verify_document.py --key <public_key> --document <document> [options]

Exit codes:
    0 - Signature valid
    1 - Signature invalid
    2 - Input error

Example:
    python verify_document.py --key keys/signer_public.key --document contract.pdf
"""

import argparse
import hashlib
import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

# Add parent directories to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "src" / "python"))

try:
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f
    )
except ImportError:
    print("Error: Could not import PQC modules. Ensure the library is installed.")
    sys.exit(1)


# Algorithm registry
ALGORITHMS = {
    "mldsa44": {"class": MLDSA44, "name": "ML-DSA-44"},
    "mldsa65": {"class": MLDSA65, "name": "ML-DSA-65"},
    "mldsa87": {"class": MLDSA87, "name": "ML-DSA-87"},
    "slh-shake-128f": {"class": SLHDSA_SHAKE_128f, "name": "SLH-DSA-SHAKE-128f"},
    "slh-shake-128s": {"class": SLHDSA_SHAKE_128s, "name": "SLH-DSA-SHAKE-128s"},
    "slh-shake-192f": {"class": SLHDSA_SHAKE_192f, "name": "SLH-DSA-SHAKE-192f"},
    "slh-shake-256f": {"class": SLHDSA_SHAKE_256f, "name": "SLH-DSA-SHAKE-256f"},
}

# Key size to algorithm mapping for auto-detection
KEY_SIZE_MAP = {
    1312: "mldsa44",
    1952: "mldsa65",
    2592: "mldsa87",
    32: "slh-shake-128f",
    48: "slh-shake-192f",
    64: "slh-shake-256f",
}


def detect_algorithm(key_data: bytes) -> str:
    """Detect algorithm from public key size."""
    key_size = len(key_data)
    if key_size in KEY_SIZE_MAP:
        return KEY_SIZE_MAP[key_size]
    raise ValueError(f"Cannot detect algorithm from key size: {key_size}")


def compute_document_hash(filepath: Path) -> tuple:
    """Compute SHA-256 and SHA-512 hashes of a document."""
    sha256_hash = hashlib.sha256()
    sha512_hash = hashlib.sha512()

    with open(filepath, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            sha256_hash.update(chunk)
            sha512_hash.update(chunk)

    return sha256_hash.hexdigest(), sha512_hash.hexdigest()


def verify_document(
    key_path: Path,
    document_path: Path,
    signature_path: Optional[Path] = None,
    quiet: bool = False,
    json_output: bool = False,
) -> dict:
    """Verify a document signature."""

    result = {
        "valid": False,
        "checks": {
            "hash": False,
            "size": False,
            "signature": False,
        },
        "document": {},
        "signer": {},
        "timestamp": None,
        "error": None,
    }

    try:
        # Read public key
        with open(key_path, "rb") as f:
            pk = f.read()

        # Default signature path
        if signature_path is None:
            signature_path = document_path.with_suffix(document_path.suffix + ".docsig")

        # Read manifest
        with open(signature_path, "r") as f:
            manifest = json.load(f)

        # Extract values from manifest
        algorithm = manifest["algorithm"]["id"]
        expected_hash = manifest["document"]["hashes"]["sha256"]
        expected_size = manifest["document"]["size"]
        signature_hex = manifest["signature"]["value"]
        context_hex = manifest["signature"].get("context", "")
        timestamp = manifest.get("timestamp", "")

        result["document"] = {
            "name": manifest["document"]["name"],
            "size": expected_size,
        }
        result["timestamp"] = timestamp
        result["algorithm"] = algorithm

        if "signer" in manifest:
            result["signer"] = manifest["signer"]

        if "signing_details" in manifest:
            result["signing_details"] = manifest["signing_details"]

        # Compute actual document hash
        actual_hash, actual_sha512 = compute_document_hash(document_path)
        actual_size = document_path.stat().st_size

        # Check hash
        result["checks"]["hash"] = (actual_hash == expected_hash)

        # Also verify SHA-512 if present
        if "sha512" in manifest["document"]["hashes"]:
            expected_sha512 = manifest["document"]["hashes"]["sha512"]
            result["checks"]["hash"] = result["checks"]["hash"] and (actual_sha512 == expected_sha512)

        # Check size
        result["checks"]["size"] = (actual_size == expected_size)

        if not result["checks"]["hash"]:
            result["error"] = "Document hash mismatch - document may have been modified"
        elif not result["checks"]["size"]:
            result["error"] = "Document size mismatch"

        # Initialize algorithm
        if algorithm not in ALGORITHMS:
            # Try to detect from key
            algorithm = detect_algorithm(pk)
        alg_info = ALGORITHMS[algorithm]
        dsa = alg_info["class"]()

        # Decode signature and context
        signature = bytes.fromhex(signature_hex)
        ctx_bytes = bytes.fromhex(context_hex) if context_hex else b"document"

        # Build message to verify (must match what was signed)
        # Extract signer info from manifest if present
        signed_data = {
            "document_hash": expected_hash,
            "document_name": manifest["document"]["name"],
            "document_size": expected_size,
            "timestamp": timestamp,
        }

        if "signer" in manifest and "name" in manifest["signer"]:
            signed_data["signer_name"] = manifest["signer"]["name"]
        if "signing_details" in manifest:
            if "reason" in manifest["signing_details"]:
                signed_data["reason"] = manifest["signing_details"]["reason"]
            if "location" in manifest["signing_details"]:
                signed_data["location"] = manifest["signing_details"]["location"]

        message = json.dumps(signed_data, sort_keys=True).encode("utf-8")

        # Verify signature
        try:
            is_valid = dsa.verify(pk, message, signature, ctx=ctx_bytes)
            result["checks"]["signature"] = is_valid
        except Exception as e:
            result["checks"]["signature"] = False
            result["error"] = f"Signature verification error: {e}"

        # Overall result
        result["valid"] = all(result["checks"].values())

        if result["valid"]:
            result["error"] = None

    except FileNotFoundError as e:
        result["error"] = str(e)
    except json.JSONDecodeError:
        result["error"] = "Invalid signature file format"
    except Exception as e:
        result["error"] = str(e)

    return result


def print_result(result: dict, quiet: bool = False, json_output: bool = False):
    """Print verification result."""
    if json_output:
        print(json.dumps(result, indent=2))
        return

    if quiet:
        return

    print()
    print("=" * 50)
    print("  Document Signature Verification")
    print("=" * 50)
    print()

    if "document" in result and result["document"]:
        print(f"Document:   {result['document'].get('name', 'Unknown')}")
        print(f"Size:       {result['document'].get('size', 0)} bytes")

    if "algorithm" in result:
        print(f"Algorithm:  {result['algorithm']}")

    if result.get("timestamp"):
        print(f"Signed:     {result['timestamp']}")

    if result.get("signer"):
        signer = result["signer"]
        if "name" in signer:
            print(f"Signer:     {signer['name']}")
        if "email" in signer:
            print(f"Email:      {signer['email']}")
        if "organization" in signer:
            print(f"Org:        {signer['organization']}")

    if result.get("signing_details"):
        details = result["signing_details"]
        if "reason" in details:
            print(f"Reason:     {details['reason']}")
        if "location" in details:
            print(f"Location:   {details['location']}")

    print()
    print("Checks:")
    checks = result.get("checks", {})
    print(f"  Hash:       {'PASS' if checks.get('hash') else 'FAIL'}")
    print(f"  Size:       {'PASS' if checks.get('size') else 'FAIL'}")
    print(f"  Signature:  {'PASS' if checks.get('signature') else 'FAIL'}")

    if result.get("error"):
        print()
        print(f"Error: {result['error']}")

    print()
    print("=" * 50)
    if result["valid"]:
        print("  SIGNATURE VALID")
    else:
        print("  SIGNATURE INVALID")
    print("=" * 50)


def main():
    parser = argparse.ArgumentParser(
        description="Post-Quantum Document Signature Verification Tool",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Exit Codes:
  0 - Signature valid
  1 - Signature invalid
  2 - Input error

Examples:
  # Verify a document signature
  %(prog)s --key keys/signer_public.key --document contract.pdf

  # JSON output for scripting
  %(prog)s --key keys/signer_public.key --document contract.pdf --json

  # Quiet mode (exit code only)
  %(prog)s --key keys/signer_public.key --document contract.pdf --quiet
        """,
    )

    parser.add_argument(
        "-k", "--key",
        required=True,
        type=Path,
        help="Public key file for verification",
    )
    parser.add_argument(
        "-d", "--document",
        required=True,
        type=Path,
        help="Document to verify",
    )
    parser.add_argument(
        "-s", "--signature",
        type=Path,
        help="Signature file (default: <document>.docsig)",
    )
    parser.add_argument(
        "-q", "--quiet",
        action="store_true",
        help="Suppress output, exit code only",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Output result as JSON",
    )

    args = parser.parse_args()

    # Validate inputs
    if not args.key.exists():
        print(f"Error: Key file not found: {args.key}", file=sys.stderr)
        sys.exit(2)
    if not args.document.exists():
        print(f"Error: Document not found: {args.document}", file=sys.stderr)
        sys.exit(2)

    result = verify_document(
        key_path=args.key,
        document_path=args.document,
        signature_path=args.signature,
        quiet=args.quiet,
        json_output=args.json,
    )

    print_result(result, quiet=args.quiet, json_output=args.json)

    sys.exit(0 if result["valid"] else 1)


if __name__ == "__main__":
    main()

File: dsa/examples/use-cases/document-signing/verify_document.cpp
/**
 * Post-Quantum Document Signature Verification Tool (C++)
 *
 * Verifies document signatures created with sign_document.
 *
 * Usage:
 *   ./verify_document <public_key> <document> [options]
 *
 * Options:
 *   --signature <file>     Signature file (default: <document>.docsig)
 *   --quiet                Exit code only
 *   --json                 Output result as JSON
 *
 * Exit codes:
 *   0 - Signature valid
 *   1 - Signature invalid
 *   2 - Input error
 *
 * Example:
 *   ./verify_document keys/signer_public.key contract.pdf
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <vector>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <regex>

#include "mldsa/mldsa.hpp"
#include "slhdsa/slh_dsa.hpp"

#include <openssl/sha.h>

// Read binary file
std::vector<uint8_t> read_file(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    return std::vector<uint8_t>(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

// Read text file
std::string read_text_file(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    return std::string(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

// Convert hex string to bytes
std::vector<uint8_t> from_hex(const std::string& hex) {
    std::vector<uint8_t> bytes;
    for (size_t i = 0; i < hex.length(); i += 2) {
        bytes.push_back(static_cast<uint8_t>(std::stoi(hex.substr(i, 2), nullptr, 16)));
    }
    return bytes;
}

// Convert bytes to hex string
std::string to_hex(const std::vector<uint8_t>& data) {
    std::ostringstream oss;
    for (uint8_t b : data) {
        oss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(b);
    }
    return oss.str();
}

// Compute SHA-256 hash
std::vector<uint8_t> sha256(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA256_DIGEST_LENGTH);
    SHA256(data.data(), data.size(), hash.data());
    return hash;
}

// Compute SHA-512 hash
std::vector<uint8_t> sha512(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA512_DIGEST_LENGTH);
    SHA512(data.data(), data.size(), hash.data());
    return hash;
}

// Extract filename from path
std::string get_filename(const std::string& path) {
    size_t pos = path.find_last_of("/\\");
    return (pos == std::string::npos) ? path : path.substr(pos + 1);
}

// Simple JSON value extraction
std::string json_get_string(const std::string& json, const std::string& key) {
    std::regex pattern("\"" + key + "\"\\s*:\\s*\"([^\"]+)\"");
    std::smatch match;
    if (std::regex_search(json, match, pattern)) {
        return match[1].str();
    }
    return "";
}

size_t json_get_int(const std::string& json, const std::string& key) {
    std::regex pattern("\"" + key + "\"\\s*:\\s*(\\d+)");
    std::smatch match;
    if (std::regex_search(json, match, pattern)) {
        return std::stoull(match[1].str());
    }
    return 0;
}

// JSON escape string
std::string json_escape(const std::string& s) {
    std::string result;
    for (char c : s) {
        switch (c) {
            case '"': result += "\\\""; break;
            case '\\': result += "\\\\"; break;
            case '\n': result += "\\n"; break;
            default: result += c;
        }
    }
    return result;
}

// Detect algorithm from public key size
std::string detect_algorithm(size_t pk_size) {
    switch (pk_size) {
        case 1312: return "mldsa44";
        case 1952: return "mldsa65";
        case 2592: return "mldsa87";
        case 32: return "slh-shake-128f";
        case 48: return "slh-shake-192f";
        case 64: return "slh-shake-256f";
        default:
            throw std::invalid_argument("Cannot detect algorithm from key size: " +
                                        std::to_string(pk_size));
    }
}

// Verify templates
template<typename DSA>
bool verify_mldsa(const std::vector<uint8_t>& pk,
                  const std::vector<uint8_t>& message,
                  const std::vector<uint8_t>& signature,
                  const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.verify(pk, message, signature, ctx);
}

template<typename DSA>
bool verify_slhdsa(const std::vector<uint8_t>& pk,
                   const std::vector<uint8_t>& message,
                   const std::vector<uint8_t>& signature,
                   const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.verify(pk, message, signature, ctx);
}

struct VerificationResult {
    bool valid = false;
    bool hash_valid = false;
    bool size_valid = false;
    bool signature_valid = false;
    std::string algorithm;
    std::string filename;
    size_t file_size = 0;
    std::string timestamp;
    std::string signer_name;
    std::string signer_email;
    std::string reason;
    std::string location;
    std::string error;
};

void print_usage() {
    std::cout << "Post-Quantum Document Signature Verification Tool (C++)\n";
    std::cout << std::string(60, '=') << "\n\n";
    std::cout << "Usage: verify_document <public_key> <document> [options]\n\n";
    std::cout << "Required:\n";
    std::cout << "  <public_key>          Public key file\n";
    std::cout << "  <document>            Document to verify\n\n";
    std::cout << "Options:\n";
    std::cout << "  --signature <file>    Signature file (default: <doc>.docsig)\n";
    std::cout << "  --quiet               Exit code only\n";
    std::cout << "  --json                Output result as JSON\n";
    std::cout << "  --help                Show this help\n\n";
    std::cout << "Exit Codes:\n";
    std::cout << "  0 - Signature valid\n";
    std::cout << "  1 - Signature invalid\n";
    std::cout << "  2 - Input error\n\n";
    std::cout << "Example:\n";
    std::cout << "  ./verify_document keys/signer_public.key contract.pdf\n";
}

void print_json_result(const VerificationResult& result) {
    std::cout << "{\n";
    std::cout << "  \"valid\": " << (result.valid ? "true" : "false") << ",\n";
    std::cout << "  \"checks\": {\n";
    std::cout << "    \"hash\": " << (result.hash_valid ? "true" : "false") << ",\n";
    std::cout << "    \"size\": " << (result.size_valid ? "true" : "false") << ",\n";
    std::cout << "    \"signature\": " << (result.signature_valid ? "true" : "false") << "\n";
    std::cout << "  },\n";
    std::cout << "  \"document\": {\n";
    std::cout << "    \"name\": \"" << json_escape(result.filename) << "\",\n";
    std::cout << "    \"size\": " << result.file_size << "\n";
    std::cout << "  },\n";
    std::cout << "  \"algorithm\": \"" << result.algorithm << "\",\n";
    std::cout << "  \"timestamp\": \"" << result.timestamp << "\"";

    if (!result.signer_name.empty()) {
        std::cout << ",\n  \"signer\": {\n";
        std::cout << "    \"name\": \"" << json_escape(result.signer_name) << "\"";
        if (!result.signer_email.empty()) {
            std::cout << ",\n    \"email\": \"" << json_escape(result.signer_email) << "\"";
        }
        std::cout << "\n  }";
    }

    if (!result.reason.empty() || !result.location.empty()) {
        std::cout << ",\n  \"signing_details\": {\n";
        bool first = true;
        if (!result.reason.empty()) {
            std::cout << "    \"reason\": \"" << json_escape(result.reason) << "\"";
            first = false;
        }
        if (!result.location.empty()) {
            if (!first) std::cout << ",\n";
            std::cout << "    \"location\": \"" << json_escape(result.location) << "\"";
        }
        std::cout << "\n  }";
    }

    if (!result.error.empty()) {
        std::cout << ",\n  \"error\": \"" << json_escape(result.error) << "\"";
    }

    std::cout << "\n}\n";
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        print_usage();
        return (argc == 1) ? 0 : 2;
    }

    std::string pk_file;
    std::string doc_file;
    std::string sig_file;
    bool quiet = false;
    bool json_output = false;

    // Parse arguments
    int i = 1;
    while (i < argc) {
        std::string arg = argv[i];

        if (arg == "--signature" && i + 1 < argc) {
            sig_file = argv[++i];
        } else if (arg == "--quiet" || arg == "-q") {
            quiet = true;
        } else if (arg == "--json") {
            json_output = true;
        } else if (arg == "--help" || arg == "-h") {
            print_usage();
            return 0;
        } else if (arg[0] != '-') {
            if (pk_file.empty()) {
                pk_file = arg;
            } else if (doc_file.empty()) {
                doc_file = arg;
            }
        } else {
            std::cerr << "Unknown option: " << arg << "\n";
            return 2;
        }
        ++i;
    }

    if (pk_file.empty() || doc_file.empty()) {
        std::cerr << "Error: Public key and document are required\n";
        return 2;
    }

    if (sig_file.empty()) {
        sig_file = doc_file + ".docsig";
    }

    VerificationResult result;

    try {
        // Read public key
        if (!quiet && !json_output) std::cerr << "Reading public key: " << pk_file << "\n";
        auto pk = read_file(pk_file);

        // Read document
        if (!quiet && !json_output) std::cerr << "Reading document: " << doc_file << "\n";
        auto doc_data = read_file(doc_file);
        result.file_size = doc_data.size();
        result.filename = get_filename(doc_file);

        // Read signature manifest
        if (!quiet && !json_output) std::cerr << "Reading signature: " << sig_file << "\n";
        std::string manifest = read_text_file(sig_file);

        // Extract values from manifest
        std::string alg_id = json_get_string(manifest, "id");
        std::string sig_hex = json_get_string(manifest, "value");
        std::string sha256_hex = json_get_string(manifest, "sha256");
        std::string sha512_hex = json_get_string(manifest, "sha512");
        std::string context_hex = json_get_string(manifest, "context");
        std::string timestamp = json_get_string(manifest, "timestamp");
        size_t expected_size = json_get_int(manifest, "size");

        // Extract signer info
        result.signer_name = json_get_string(manifest, "name");
        result.signer_email = json_get_string(manifest, "email");
        result.reason = json_get_string(manifest, "reason");
        result.location = json_get_string(manifest, "location");

        result.algorithm = alg_id;
        result.timestamp = timestamp;

        // Detect algorithm from key if not in manifest
        if (alg_id.empty()) {
            alg_id = detect_algorithm(pk.size());
            result.algorithm = alg_id;
        }

        // Verify hashes
        auto computed_sha256 = sha256(doc_data);
        auto computed_sha512 = sha512(doc_data);
        std::string computed_sha256_hex = to_hex(computed_sha256);
        std::string computed_sha512_hex = to_hex(computed_sha512);

        result.hash_valid = (computed_sha256_hex == sha256_hex);
        if (!sha512_hex.empty()) {
            result.hash_valid = result.hash_valid && (computed_sha512_hex == sha512_hex);
        }

        // Verify size
        result.size_valid = (doc_data.size() == expected_size);

        if (!result.hash_valid) {
            result.error = "Document hash mismatch - document may have been modified";
        } else if (!result.size_valid) {
            result.error = "Document size mismatch";
        }

        // Build message to verify
        std::string doc_name = json_get_string(manifest, "name");
        if (doc_name.empty()) doc_name = result.filename;

        std::ostringstream msg_stream;
        msg_stream << "{\"document_hash\":\"" << sha256_hex << "\",";
        msg_stream << "\"document_name\":\"" << doc_name << "\",";
        msg_stream << "\"document_size\":" << expected_size << ",";
        msg_stream << "\"timestamp\":\"" << timestamp << "\"";
        if (!result.signer_name.empty()) {
            msg_stream << ",\"signer_name\":\"" << result.signer_name << "\"";
        }
        if (!result.reason.empty()) {
            msg_stream << ",\"reason\":\"" << result.reason << "\"";
        }
        if (!result.location.empty()) {
            msg_stream << ",\"location\":\"" << result.location << "\"";
        }
        msg_stream << "}";
        std::string msg_str = msg_stream.str();
        std::vector<uint8_t> message(msg_str.begin(), msg_str.end());

        // Decode signature and context
        std::vector<uint8_t> signature = from_hex(sig_hex);
        std::vector<uint8_t> ctx = from_hex(context_hex);

        // Verify signature
        if (!quiet && !json_output) std::cerr << "Verifying signature with " << alg_id << "...\n";

        if (alg_id == "mldsa44") {
            result.signature_valid = verify_mldsa<mldsa::MLDSA44>(pk, message, signature, ctx);
        } else if (alg_id == "mldsa65") {
            result.signature_valid = verify_mldsa<mldsa::MLDSA65>(pk, message, signature, ctx);
        } else if (alg_id == "mldsa87") {
            result.signature_valid = verify_mldsa<mldsa::MLDSA87>(pk, message, signature, ctx);
        } else if (alg_id == "slh-shake-128f") {
            result.signature_valid = verify_slhdsa<slhdsa::SLHDSA_SHAKE_128f>(pk, message, signature, ctx);
        } else if (alg_id == "slh-shake-192f") {
            result.signature_valid = verify_slhdsa<slhdsa::SLHDSA_SHAKE_192f>(pk, message, signature, ctx);
        } else if (alg_id == "slh-shake-256f") {
            result.signature_valid = verify_slhdsa<slhdsa::SLHDSA_SHAKE_256f>(pk, message, signature, ctx);
        } else {
            throw std::invalid_argument("Unsupported algorithm: " + alg_id);
        }

        // Final result
        result.valid = result.hash_valid && result.size_valid && result.signature_valid;

    } catch (const std::exception& e) {
        result.error = e.what();
        if (json_output) {
            print_json_result(result);
        } else if (!quiet) {
            std::cerr << "Error: " << e.what() << "\n";
        }
        return 2;
    }

    // Output result
    if (json_output) {
        print_json_result(result);
    } else if (!quiet) {
        std::cout << "\n" << std::string(50, '=') << "\n";
        std::cout << "  Document Signature Verification\n";
        std::cout << std::string(50, '=') << "\n\n";

        std::cout << "Document:   " << result.filename << "\n";
        std::cout << "Size:       " << result.file_size << " bytes\n";
        std::cout << "Algorithm:  " << result.algorithm << "\n";
        std::cout << "Timestamp:  " << result.timestamp << "\n";

        if (!result.signer_name.empty()) {
            std::cout << "Signer:     " << result.signer_name << "\n";
        }
        if (!result.signer_email.empty()) {
            std::cout << "Email:      " << result.signer_email << "\n";
        }
        if (!result.reason.empty()) {
            std::cout << "Reason:     " << result.reason << "\n";
        }
        if (!result.location.empty()) {
            std::cout << "Location:   " << result.location << "\n";
        }

        std::cout << "\nChecks:\n";
        std::cout << "  Hash:       " << (result.hash_valid ? "PASS" : "FAIL") << "\n";
        std::cout << "  Size:       " << (result.size_valid ? "PASS" : "FAIL") << "\n";
        std::cout << "  Signature:  " << (result.signature_valid ? "PASS" : "FAIL") << "\n";

        if (!result.error.empty()) {
            std::cout << "\nError: " << result.error << "\n";
        }

        std::cout << "\n" << std::string(50, '=') << "\n";
        if (result.valid) {
            std::cout << "  SIGNATURE VALID\n";
        } else {
            std::cout << "  SIGNATURE INVALID\n";
        }
        std::cout << std::string(50, '=') << "\n";
    }

    return result.valid ? 0 : 1;
}
